File: python/tracing/exa/04_expected_error.py
from __future__ import annotations

from _shared import clients, example_attributes, finish, make_respan
from respan import workflow

EXAMPLE = "expected-error"


@workflow(name="exa_python_expected_error")
def expected_error(client) -> str:
    try:
        client.search("expected Exa provider error", num_results=1)
    except Exception as exc:  # noqa: BLE001 - this scenario validates error spans.
        return f"{type(exc).__name__}: {exc}"
    raise AssertionError("expected the deterministic Exa request to fail")


def main() -> None:
    respan = make_respan(EXAMPLE)
    try:
        with clients() as (client, _async_client, mode):
            if mode == "live":
                print("Skipping destructive expected-error request in live mode.")
                return
            with example_attributes(EXAMPLE, mode) as workflow_name:
                result = expected_error(client)
                print({"workflow_name": workflow_name, "mode": mode, "error": result})
    finally:
        finish(respan)


if __name__ == "__main__":
    main()

File: typescript/tracing/exa/04_expected_error.ts
import { printResult, runExaExample } from "./_shared.js";

const example = "expected-error";
const result = await runExaExample({
  example,
  fn: async ({ client, mode }) => {
    if (mode === "live") return { mode, skipped: true };
    try {
      await client.search("expected Exa provider error", { numResults: 1 });
    } catch (error) {
      return {
        mode,
        error: error instanceof Error ? `${error.name}: ${error.message}` : String(error),
      };
    }
    throw new Error("expected the deterministic Exa request to fail");
  },
});
printResult(example, result.mode, result);
